sdk/Trace/SpanBuilder.php

```diff
@@ -0,0 +1,212 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Sdk\Trace;
+
+use function in_array;
+use OpenTelemetry\Context\Context;
+use OpenTelemetry\Sdk\InstrumentationLibrary;
+use OpenTelemetry\Trace as API;
+use OpenTelemetry\Trace\SpanKind; /** @phan-suppress-current-line PhanUnreferencedUseNormal */
+
+final class SpanBuilder implements API\SpanBuilder
+{
+    /**
+     * @var non-empty-string
+     * @readonly
+     */
+    private string $spanName;
+
+    /** @readonly */
+    private InstrumentationLibrary $instrumentationLibrary;
+
+    /** @readonly */
+    private TracerSharedState $tracerSharedState;
+
+    /** @readonly */
+    private SpanLimits $spanLimits;
+
+    private ?Context $parentContext = null; // Null means use current context.
+
+    /**
+     * @psalm-var API\SpanKind::KIND_*
+     */
+    private int $spanKind = API\SpanKind::KIND_INTERNAL;
+    private ?API\Attributes $attributes = null;
+    private ?API\Links $links = null;
+    private int $totalNumberOfLinksAdded = 0;
+    private int $startEpochNanos = 0;
+
+    /** @param non-empty-string $spanName */
+    public function __construct(
+        string $spanName,
+        InstrumentationLibrary $instrumentationLibrary,
+        TracerSharedState $tracerSharedState,
+        SpanLimits $spanLimits
+    ) {
+        $this->spanName = $spanName;
+        $this->instrumentationLibrary = $instrumentationLibrary;
+        $this->tracerSharedState = $tracerSharedState;
+        $this->spanLimits = $spanLimits;
+    }
+
+    /** @inheritDoc */
+    public function setParent(Context $parentContext): API\SpanBuilder
+    {
+        $this->parentContext = $parentContext;
+
+        return $this;
+    }
+
+    /** @inheritDoc */
+    public function setNoParent(): API\SpanBuilder
+    {
+        $this->parentContext = Context::getRoot();
+
+        return $this;
+    }
+
+    /** @inheritDoc */
+    public function addLink(API\SpanContext $context, API\Attributes $attributes = null): API\SpanBuilder
+    {
+        if (!$context->isValid()) {
+            return $this;
+        }
+
+        $this->totalNumberOfLinksAdded++;
+
+        if (null === $this->links) {
+            $this->links = new Links();
+        }
+
+        if (count($this->links) === $this->spanLimits->getLinkCountLimit()) {
+            return $this;
+        }
+
+        $this->links->addLink(new Link($context, $attributes));
+
+        return $this;
+    }
+
+    /** @inheritDoc */
+    public function setAttribute(string $key, $value): API\SpanBuilder
```

Comment:
The `SpanLimits` are missed here. As soon as we have `$this->spanLimits` set in the constructor we may use it.
`Span` applies the limits only when initializes an empty `Attributes` collection: https://github.com/open-telemetry/opentelemetry-php/blob/189166e477a065520e438a7a705b28650595de9c/sdk/Trace/Span.php#L269-L271
Some tests on `SpanBuilder::setAttribute` could help to cover this issue.

Reply:
Thanks! I handled and added tests for both of these cases. See latest commit.